File: Course_1-Algorithmic_Toolbox/Week-4/Excercise_Challenges/3_improving_quicksort/sorting.cpp
#include <iostream>
#include <vector>
#include <cstdlib>
using namespace std;

int partition2(vector<int> &a, int l, int r)
{
    int x = a[l];
    int j = l;
    for(int i = l + 1; i <= r; i++){
        if(a[i] <= x){
            j++;
            swap(a[i], a[j]);
        }
    }
    swap(a[l], a[j]);
    return j;
}

vector<int> partition3(vector<int> &a, int l, int r)
{
    vector<int> m(2);
    int pivot = a[l];
    int mid = l;

    while(mid <= r){
        if(a[mid] < pivot){
            swap(a[mid++], a[l++]);
        }
        else if(a[mid] == pivot){
            mid++;
        }
        else if(a[mid] > pivot){
            swap(a[mid], a[r--]);
        }
    }
    m[0] = l;
    m[1] = r;

    return m;
}

void randomized_quick_sort(vector<int> &a, int l, int r)
{
    if(l >= r){
        return;
    }

    int k = l + rand() % (r - l + 1);
    swap(a[l], a[k]);
    vector<int> m = partition3(a, l, r);

    randomized_quick_sort(a, l, m[0] - 1);
    randomized_quick_sort(a, m[1] + 1, r);
}

int main(void)
{
    int n;
    cin >> n;
  
    vector<int> a(n);
    for(size_t i = 0; i < a.size(); ++i){
        cin >> a[i];
    }

    randomized_quick_sort(a, 0, a.size() - 1);
    for(size_t i = 0; i < a.size(); ++i){
        cout << a[i] << ' ';
    }

    return 0;
}

File: Course_1-Algorithmic_Toolbox/Week-4/Excercise_Challenges/3_improving_quicksort/sorting.py
import random

def partition3(a, l, r):
    pivot = a[l]
    mid = l

    while mid <= r:
        if a[mid] < pivot:
            a[mid], a[l] = a[l], a[mid]
            mid += 1
            l += 1
        elif a[mid] == pivot:
            mid += 1
        else:
            a[mid], a[r] = a[r], a[mid]
            r -= 1
    
    i, j = l, r
    return i, j

def partition2(a, l, r):
    x = a[l]
    j = l
    for i in range(l + 1, r + 1):
        if a[i] <= x:
            j += 1
            a[i], a[j] = a[j], a[i]
    a[l], a[j] = a[j], a[l]
    return j


def randomized_quick_sort(a, l, r):
    if l >= r:
        return
    k = random.randint(l, r)
    a[l], a[k] = a[k], a[l]

    i, j = partition3(a, l, r)
    randomized_quick_sort(a, l, i - 1)
    randomized_quick_sort(a, j + 1, r)


if __name__ == '__main__':
    n = int(input())
    a = list(map(int, input().split()))

    randomized_quick_sort(a, 0, n - 1)
    for x in a:
        print(x, end=' ')
